crates/ra_hir/src/code_model.rs

```diff
@@ -469,7 +469,7 @@ impl EnumVariant {
         self.variant_data(db).kind()
     }
 
-    pub(crate) fn variant_data(self, db: &dyn HirDatabase) -> Arc<VariantData> {
+    pub fn variant_data(self, db: &dyn HirDatabase) -> Arc<VariantData> {
```

Comment:
Do we want to make this public? I think this PR could have been done using the existing `fields` and `kind` methods, but matching on `VariantData` has nicer ergonomics IMO.

Reply:
Good question. The current thinking is that, no, we don't want to do that. It's unclear that this thinking is right, but it's better to stick to it until we figure out that it's definitely wrong. 

Generally, the API is split between internal `hir_xxx` crates, and the facade `hir` crate. Types in `hir`, on the first blush, stupidly duplicate types in `hir_xxx`. However, there are two main reasons for that.

*First*, `hir_xxx` observe the rust-code in process of analysis, while `hir` provides a conceptually "complete" view of code. So, for example, `VartiantData` does not include information about types, because types appear later in the pipeline. However, `EnumVariant::fields` returns fields which "know" their type. 

*Second*, `hir_xxx` types generally don't include parent pointers. They use relative IDs, and require the users to maintain the stack of "this is nth field of kth variant of this enum". `hir` types include parent IDs and, eg, having `hir::StructField` and a `db` is enough to get anything you could possible get.

There's also a practical consideration that, hopefully, `hir` layer should allow us to do crazy refactorings to the way we represent various datums, without changing the IDE code. 

So what I'd suggest here is to keep the method private and either use existing pair of `kind/fields` methods, or to change signature of `.kind` to return

```rust
enum hir::StructKind {
  Unit, Tuple(Vec<hir::StructField>), Record(Vec<hir::StructField>)
}
```

Reply:
Also, a better terminology is need, as we sometimes use `Variant`, and sometimes `Struct`, and it is confusing... A bad option is `Product`.

Reply:
Hmm, do we use `Struct` anywhere where it doesn't refer to `struct` in particular? The way I think of it is that "a thing with fields" is a `Variant`, an `enum` has multiple of those, and a `struct` just one.

Reply:
Ok, I guess `StructField` and `StructKind` are the cases you mean :sweat_smile:

Reply:
+1 for field. Not entirely sure about variant -- it is better than struct, but it still feels funny to me that a struct has a variant. IE, variant is biased towards enums.

Reply:
Thanks for the detailed explanation. I created a new `hir::StructKind` type, and export that type now from `hir` rather than exporting the `hir_def::StructKind` (which doesn't contain the fields). Is this what you intended? 

It seems one downside to this is that now the IDE completion code has to wait for the fields to be resolved, even though it doesn't use them. I'm not sure how much of a performance impact something like this has, so I don't know if this is something that is okay, or should be avoided?

Reply:
>I don't know if this is something that is okay, or should be avoided?

That should be OK for two reasons:

* generally, the IDE-level code is cold-ish, as it only deals with things on the screen, and there's a constant amount of those
* I think we should not compromise API's convenience in the name of performance here. It might be the case that we'll need something more lazy here eventually, but the laziness should be hidden in the API. The client should work as if all the info is available at all times.